File: data/src/main/java/org/cryptomator/data/cloud/webdav/network/WebDavClient.java
package org.cryptomator.data.cloud.webdav.network;

import static java.util.Collections.sort;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import org.cryptomator.data.cloud.webdav.WebDavFolder;
import org.cryptomator.data.cloud.webdav.WebDavNode;
import org.cryptomator.domain.CloudFolder;
import org.cryptomator.domain.CloudNode;
import org.cryptomator.domain.exception.AlreadyExistException;
import org.cryptomator.domain.exception.BackendException;
import org.cryptomator.domain.exception.CloudNodeAlreadyExistsException;
import org.cryptomator.domain.exception.FatalBackendException;
import org.cryptomator.domain.exception.ForbiddenException;
import org.cryptomator.domain.exception.NotFoundException;
import org.cryptomator.domain.exception.ParentFolderDoesNotExistException;
import org.cryptomator.domain.exception.ServerNotWebdavCompatibleException;
import org.cryptomator.domain.exception.TypeMismatchException;
import org.cryptomator.domain.exception.UnauthorizedException;
import org.cryptomator.domain.usecases.cloud.DataSource;
import org.xmlpull.v1.XmlPullParserException;

import android.content.Context;

import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

class WebDavClient {

	private final Context context;
	private final WebDavCompatibleHttpClient httpClient;

	WebDavClient(Context context, WebDavCompatibleHttpClient httpClient) {
		this.context = context;
		this.httpClient = httpClient;
	}

	List<CloudNode> dirList(String url, WebDavFolder listedFolder) throws BackendException {
		try (Response response = executePropfindRequest(url, PropfindDepth.ONE)) {
			checkPropfindExecutionSucceeded(response.code());

			List<PropfindEntryData> nodes = getEntriesFromResponse(listedFolder, response);

			return processDirList(nodes, listedFolder);
		} catch (IOException | XmlPullParserException e) {
			throw new FatalBackendException(e);
		}
	}

	public WebDavNode get(String url, CloudFolder parent) throws BackendException {
		try (Response response = executePropfindRequest(url, PropfindDepth.ZERO)) {
			checkPropfindExecutionSucceeded(response.code());

			List<PropfindEntryData> nodes = getEntriesFromResponse((WebDavFolder) parent, response);

			return processGet(nodes, (WebDavFolder) parent);
		} catch (IOException | XmlPullParserException e) {
			throw new FatalBackendException(e);
		}
	}

	private Response executePropfindRequest(String url, PropfindDepth depth) throws IOException {
		String body = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n" //
				+ "<d:propfind xmlns:d=\"DAV:\">\n" //
				+ "<d:prop>\n" //
				+ "<d:resourcetype />\n" //
				+ "<d:getcontentlength />\n" //
				+ "<d:getlastmodified />\n" //
				+ "</d:prop>\n" //
				+ "</d:propfind>";

		Request.Builder builder = new Request.Builder() //
				.method("PROPFIND", RequestBody.create(MediaType.parse(body), body)) //
				.url(url) //
				.header("DEPTH", depth.value) //
				.header("Content-Type", "text/xml");

		return httpClient.execute(builder);
	}

	private void checkPropfindExecutionSucceeded(int responseCode) throws BackendException {
		switch (responseCode) {
		case HttpURLConnection.HTTP_UNAUTHORIZED:
			throw new UnauthorizedException();
		case HttpURLConnection.HTTP_FORBIDDEN:
			throw new ForbiddenException();
		case HttpURLConnection.HTTP_NOT_FOUND:
			throw new NotFoundException();
		}

		if (responseCode < 199 || responseCode > 300) {
			throw new FatalBackendException("Response code isn't between 200 and 300: " + responseCode);
		}
	}

	private List<PropfindEntryData> getEntriesFromResponse(WebDavFolder listedFolder, Response response) throws IOException, XmlPullParserException {
		try (final ResponseBody responseBody = response.body()) {
			return new PropfindResponseParser(listedFolder).parse(responseBody.byteStream());
		}
	}

	public void move(String from, String to) throws BackendException {
		Request.Builder builder = new Request.Builder() //
				.method("MOVE", null) //
				.url(from) //
				.header("Content-Type", "text/xml") //
				.header("Destination", to) //
				.header("Depth", "infinity") //
				.header("Overwrite", "F");

		try (Response response = httpClient.execute(builder)) {
			if (!response.isSuccessful()) {
				switch (response.code()) {
				case HttpURLConnection.HTTP_UNAUTHORIZED:
					throw new UnauthorizedException();
				case HttpURLConnection.HTTP_FORBIDDEN:
					throw new ForbiddenException();
				case HttpURLConnection.HTTP_NOT_FOUND:
					throw new NotFoundException();
				case HttpURLConnection.HTTP_CONFLICT:
					throw new ParentFolderDoesNotExistException();
				case HttpURLConnection.HTTP_PRECON_FAILED:
					throw new CloudNodeAlreadyExistsException(to);
				default:
					throw new FatalBackendException("Response code isn't between 200 and 300: " + response.code());
				}
			}
		} catch (IOException e) {
			throw new FatalBackendException(e);
		}
	}

	InputStream readFile(String url) throws BackendException {
		Request.Builder builder = new Request.Builder() //
				.get() //
				.url(url);

		Response response = null;
		boolean success = false;

		try {
			response = httpClient.execute(builder);

			if (response.isSuccessful()) {
				success = true;
				return response.body().byteStream();
			} else {
				switch (response.code()) {
				case HttpURLConnection.HTTP_UNAUTHORIZED:
					throw new UnauthorizedException();
				case HttpURLConnection.HTTP_FORBIDDEN:
					throw new ForbiddenException();
				case HttpURLConnection.HTTP_NOT_FOUND:
					throw new NotFoundException();
				case 416: // UNSATISFIABLE_RANGE
					return new ByteArrayInputStream(new byte[0]);
				default:
					throw new FatalBackendException("Response code isn't between 200 and 300: " + response.code());
				}
			}
		} catch (IOException e) {
			throw new FatalBackendException(e);
		} finally {
			if (response != null && !success) {
				response.close();
			}
		}
	}

	void writeFile(String url, DataSource data) throws BackendException {
		Request.Builder builder = new Request.Builder() //
				.put(DataSourceBasedRequestBody.from(context, data)) //
				.url(url);

		try (Response response = httpClient.execute(builder)) {
			if (!response.isSuccessful()) {
				switch (response.code()) {
				case HttpURLConnection.HTTP_UNAUTHORIZED:
					throw new UnauthorizedException();
				case HttpURLConnection.HTTP_FORBIDDEN:
					throw new ForbiddenException();
				case HttpURLConnection.HTTP_BAD_METHOD:
					throw new TypeMismatchException();
				case HttpURLConnection.HTTP_CONFLICT: // fall through
				case HttpURLConnection.HTTP_NOT_FOUND: // necessary due to a bug in Nextcloud, see https://github.com/nextcloud/server/issues/23519
					throw new ParentFolderDoesNotExistException();
				default:
					throw new FatalBackendException("Response code isn't between 200 and 300: " + response.code());
				}
			}
		} catch (IOException e) {
			throw new FatalBackendException(e);
		}
	}

	WebDavFolder createFolder(String path, WebDavFolder folder) throws BackendException {
		Request.Builder builder = new Request.Builder() //
				.method("MKCOL", null) //
				.url(path);

		try (Response response = httpClient.execute(builder)) {
			if (response.isSuccessful()) {
				return folder;
			} else {
				switch (response.code()) {
				case HttpURLConnection.HTTP_UNAUTHORIZED:
					throw new UnauthorizedException();
				case HttpURLConnection.HTTP_FORBIDDEN:
					throw new ForbiddenException();
				case HttpURLConnection.HTTP_BAD_METHOD:
					throw new AlreadyExistException();
				case HttpURLConnection.HTTP_CONFLICT:
					throw new ParentFolderDoesNotExistException();
				default:
					throw new FatalBackendException("Response code isn't between 200 and 300: " + response.code());
				}
			}
		} catch (IOException e) {
			throw new FatalBackendException(e);
		}
	}

	public void delete(String url) throws BackendException {
		Request.Builder builder = new Request.Builder() //
				.delete() //
				.url(url);

		try (Response response = httpClient.execute(builder)) {
			if (!response.isSuccessful()) {
				switch (response.code()) {
				case HttpURLConnection.HTTP_UNAUTHORIZED:
					throw new UnauthorizedException();
				case HttpURLConnection.HTTP_FORBIDDEN:
					throw new ForbiddenException();
				case HttpURLConnection.HTTP_NOT_FOUND:
					throw new NotFoundException(String.format("Node %s doesn't exists", url));
				default:
					throw new FatalBackendException("Response code isn't between 200 and 300: " + response.code());
				}
			}
		} catch (IOException e) {
			throw new FatalBackendException(e);
		}
	}

	void checkAuthenticationAndServerCompatibility(String url) throws BackendException {
		final Request.Builder optionsRequest = new Request.Builder() //
				.method("OPTIONS", null) //
				.url(url);

		try (Response response = httpClient.execute(optionsRequest)) {
			if (response.isSuccessful()) {
				final boolean containsDavHeader = response.headers().names().contains("DAV");
				if (!containsDavHeader) {
					throw new ServerNotWebdavCompatibleException();
				}
			} else {
				switch (response.code()) {
				case HttpURLConnection.HTTP_UNAUTHORIZED:
					throw new UnauthorizedException();
				case HttpURLConnection.HTTP_FORBIDDEN:
					throw new ForbiddenException();
				default:
					throw new FatalBackendException("Response code isn't between 200 and 300: " + response.code());
				}
			}
		} catch (IOException e) {
			throw new FatalBackendException(e);
		}

		try (Response response = executePropfindRequest(url, PropfindDepth.ZERO)) {
			checkPropfindExecutionSucceeded(response.code());
		} catch (IOException e) {
			throw new FatalBackendException(e);
		}
	}

	private List<CloudNode> processDirList(List<PropfindEntryData> entryData, WebDavFolder requestedFolder) {
		List<CloudNode> result = new ArrayList<>();
		sort(entryData, ASCENDING_BY_DEPTH);
		// after sorting the first entry is the parent
		// because it's depth is 1 smaller than the depth
		// ot the other entries, thus we skip the first entry
		for (PropfindEntryData childEntry : entryData.subList(1, entryData.size())) {
			result.add(childEntry.toCloudNode(requestedFolder));
		}
		return result;
	}

	private WebDavNode processGet(List<PropfindEntryData> entryData, WebDavFolder requestedFolder) {
		sort(entryData, ASCENDING_BY_DEPTH);
		return entryData.size() >= 1 ? entryData.get(0).toCloudNode(requestedFolder) : null;
	}

	private final Comparator<PropfindEntryData> ASCENDING_BY_DEPTH = (o1, o2) -> o1.getDepth() - o2.getDepth();

	private enum PropfindDepth {
		ZERO("0"), //
		ONE("1"), //
		INFINITY("infinity");

		private final String value;

		PropfindDepth(final String value) {
			this.value = value;
		}
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/webdav/network/WebDavClient.kt
package org.cryptomator.data.cloud.webdav.network

import org.cryptomator.data.cloud.webdav.WebDavFolder
import org.cryptomator.data.cloud.webdav.WebDavNode
import org.cryptomator.domain.CloudFolder
import org.cryptomator.domain.exception.AlreadyExistException
import org.cryptomator.domain.exception.BackendException
import org.cryptomator.domain.exception.CloudNodeAlreadyExistsException
import org.cryptomator.domain.exception.FatalBackendException
import org.cryptomator.domain.exception.ForbiddenException
import org.cryptomator.domain.exception.NotFoundException
import org.cryptomator.domain.exception.ParentFolderDoesNotExistException
import org.cryptomator.domain.exception.TypeMismatchException
import org.cryptomator.domain.exception.UnauthorizedException
import org.cryptomator.util.Optional
import org.xmlpull.v1.XmlPullParserException
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.util.Collections
import java.util.Date
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

internal class WebDavClient(private val httpClient: WebDavCompatibleHttpClient) {

	private val ASCENDING_BY_DEPTH = Comparator { o1: PropfindEntryData, o2: PropfindEntryData -> o1.getDepth() - o2.getDepth() }

	@Throws(BackendException::class)
	fun dirList(url: String, listedFolder: WebDavFolder): List<WebDavNode> {
		try {
			executePropfindRequest(url, PropfindDepth.ONE).use { response ->
				checkPropfindExecutionSucceeded(response.code)
				val nodes = getEntriesFromResponse(listedFolder, response)
				return processDirList(nodes, listedFolder)
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		} catch (e: XmlPullParserException) {
			throw FatalBackendException(e)
		}
	}

	@Throws(BackendException::class)
	operator fun get(url: String, parent: CloudFolder): WebDavNode? {
		try {
			executePropfindRequest(url, PropfindDepth.ZERO).use { response ->
				checkPropfindExecutionSucceeded(response.code)
				val nodes = getEntriesFromResponse(parent as WebDavFolder, response)
				return processGet(nodes, parent)
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		} catch (e: XmlPullParserException) {
			throw FatalBackendException(e)
		}
	}

	@Throws(IOException::class)
	private fun executePropfindRequest(url: String, depth: PropfindDepth): Response {
		val body = """<?xml version="1.0" encoding="utf-8"?>
						<d:propfind xmlns:d="DAV:">
						<d:prop>
						<d:resourcetype />
						<d:getcontentlength />
						<d:getlastmodified />
						</d:prop>
						</d:propfind>"""
		val builder = Request.Builder() //
			.method("PROPFIND", body.toRequestBody(body.toMediaTypeOrNull())) //
			.url(url) //
			.header("DEPTH", depth.value) //
			.header("Content-Type", "text/xml")
		return httpClient.execute(builder)
	}

	@Throws(BackendException::class)
	private fun checkPropfindExecutionSucceeded(responseCode: Int) {
		when (responseCode) {
			HttpURLConnection.HTTP_UNAUTHORIZED -> throw UnauthorizedException()
			HttpURLConnection.HTTP_FORBIDDEN -> throw ForbiddenException()
			HttpURLConnection.HTTP_NOT_FOUND -> throw NotFoundException()
		}
		if (responseCode < 199 || responseCode > 300) {
			throw FatalBackendException("Response code isn't between 200 and 300: $responseCode")
		}
	}

	@Throws(IOException::class, XmlPullParserException::class)
	private fun getEntriesFromResponse(listedFolder: WebDavFolder, response: Response): List<PropfindEntryData> {
		return response.body?.use { responseBody -> return PropfindResponseParser(listedFolder).parse(responseBody.byteStream()) } ?: emptyList()
	}

	@Throws(BackendException::class)
	fun move(from: String, to: String) {
		val builder = Request.Builder() //
			.method("MOVE", null) //
			.url(from) //
			.header("Content-Type", "text/xml") //
			.header("Destination", to) //
			.header("Depth", "infinity") //
			.header("Overwrite", "F")
		try {
			httpClient.execute(builder).use { response ->
				if (!response.isSuccessful) {
					when (response.code) {
						HttpURLConnection.HTTP_UNAUTHORIZED -> throw UnauthorizedException()
						HttpURLConnection.HTTP_FORBIDDEN -> throw ForbiddenException()
						HttpURLConnection.HTTP_NOT_FOUND -> throw NotFoundException()
						HttpURLConnection.HTTP_CONFLICT -> throw ParentFolderDoesNotExistException()
						HttpURLConnection.HTTP_PRECON_FAILED -> throw CloudNodeAlreadyExistsException(to)
						else -> throw FatalBackendException("Response code isn't between 200 and 300: " + response.code)
					}
				}
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		}
	}

	@Throws(BackendException::class)
	fun readFile(url: String): InputStream {
		val builder = Request.Builder() //
			.get() //
			.url(url)
		var response: Response? = null
		var success = false
		return try {
			response = httpClient.execute(builder)
			if (response.isSuccessful) {
				success = true
				response.body?.byteStream() ?: throw FatalBackendException("Response body is null")
			} else {
				when (response.code) {
					HttpURLConnection.HTTP_UNAUTHORIZED -> throw UnauthorizedException()
					HttpURLConnection.HTTP_FORBIDDEN -> throw ForbiddenException()
					HttpURLConnection.HTTP_NOT_FOUND -> throw NotFoundException()
					416 -> ByteArrayInputStream(ByteArray(0))
					else -> throw FatalBackendException("Response code isn't between 200 and 300: " + response.code)
				}
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		} finally {
			if (response != null && !success) {
				response.close()
			}
		}
	}

	@Throws(BackendException::class)
	fun writeFile(url: String, inputStream: InputStream, modifiedDate: Date) {
		val builder = Request.Builder() //
			.addHeader("X-OC-Mtime", modifiedDate.toInstant().toEpochMilli().div(1000).toString()) //
			.put(InputStreamSourceBasedRequestBody.from(inputStream)) //
			.url(url)
		try {
			httpClient.execute(builder).use { response ->
				if (!response.isSuccessful) {
					when (response.code) {
						HttpURLConnection.HTTP_UNAUTHORIZED -> throw UnauthorizedException()
						HttpURLConnection.HTTP_FORBIDDEN -> throw ForbiddenException()
						HttpURLConnection.HTTP_BAD_METHOD -> throw TypeMismatchException()
						HttpURLConnection.HTTP_CONFLICT, HttpURLConnection.HTTP_NOT_FOUND -> throw ParentFolderDoesNotExistException()
						else -> throw FatalBackendException("Response code isn't between 200 and 300: " + response.code)
					}
				}
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		}
	}

	@Throws(BackendException::class)
	fun createFolder(path: String, folder: WebDavFolder): WebDavFolder {
		val builder = Request.Builder() //
			.method("MKCOL", null) //
			.url(path)
		try {
			httpClient.execute(builder).use { response ->
				return if (response.isSuccessful) {
					folder
				} else {
					when (response.code) {
						HttpURLConnection.HTTP_UNAUTHORIZED -> throw UnauthorizedException()
						HttpURLConnection.HTTP_FORBIDDEN -> throw ForbiddenException()
						HttpURLConnection.HTTP_BAD_METHOD -> throw AlreadyExistException()
						HttpURLConnection.HTTP_CONFLICT -> throw ParentFolderDoesNotExistException()
						else -> throw FatalBackendException("Response code isn't between 200 and 300: " + response.code)
					}
				}
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		}
	}

	@Throws(BackendException::class)
	fun delete(url: String) {
		val builder = Request.Builder() //
			.delete() //
			.url(url)
		try {
			httpClient.execute(builder).use { response ->
				if (!response.isSuccessful) {
					when (response.code) {
						HttpURLConnection.HTTP_UNAUTHORIZED -> throw UnauthorizedException()
						HttpURLConnection.HTTP_FORBIDDEN -> throw ForbiddenException()
						HttpURLConnection.HTTP_NOT_FOUND -> throw NotFoundException(String.format("Node %s doesn't exists", url))
						else -> throw FatalBackendException("Response code isn't between 200 and 300: " + response.code)
					}
				}
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		}
	}

	@Throws(BackendException::class)
	fun checkAuthenticationAndServerCompatibility(url: String) {
		val optionsRequest = Request.Builder() //
			.method("OPTIONS", null) //
			.url(url)
		try {
			httpClient.execute(optionsRequest).use { response ->
				if (response.isSuccessful) {
					val containsDavHeader = response.headers.names().contains("DAV")
					if (!containsDavHeader) {
						throw ServerNotWebdavCompatibleException()
					}
				} else {
					when (response.code) {
						HttpURLConnection.HTTP_UNAUTHORIZED -> throw UnauthorizedException()
						HttpURLConnection.HTTP_FORBIDDEN -> throw ForbiddenException()
						else -> throw FatalBackendException("Response code isn't between 200 and 300: " + response.code)
					}
				}
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		}
		try {
			executePropfindRequest(url, PropfindDepth.ZERO).use { response -> checkPropfindExecutionSucceeded(response.code) }
		} catch (e: IOException) {
			throw FatalBackendException(e)
		}
	}

	private fun processDirList(entryData: List<PropfindEntryData>, requestedFolder: WebDavFolder): List<WebDavNode> {
		Collections.sort(entryData, ASCENDING_BY_DEPTH)
		// after sorting the first entry is the parent
		// because it's depth is 1 smaller than the depth
		// ot the other entries, thus we skip the first entry
		return entryData.subList(1, entryData.size).mapTo(ArrayList()) { it.toCloudNode(requestedFolder) }
	}

	private fun processGet(entryData: List<PropfindEntryData>, requestedFolder: WebDavFolder): WebDavNode? {
		Collections.sort(entryData, ASCENDING_BY_DEPTH)
		return if (entryData.isNotEmpty()) entryData[0].toCloudNode(requestedFolder) else null
	}

	private enum class PropfindDepth(val value: String) {
		ZERO("0"),  //
		ONE("1"),  //
		INFINITY("infinity");
	}
}
